In [2]:
!pip install python-dotenv pandas numpy openpyxl sqlalchemy

error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.
    
    See /usr/share/doc/python3.12/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detai

In [3]:
import os
from dotenv import load_dotenv

load_dotenv()

DB_HOST = os.getenv("SANDHYA_ERP_DB_HOST")
DB_PORT = os.getenv("SANDHYA_ERP_DB_PORT")

DB_USERNAME = os.getenv("SANDHYA_ERP_DB_USERNAME")
DB_PASSWORD = os.getenv("SANDHYA_ERP_DB_PASSWORD")
DB_NAME = os.getenv("SANDHYA_ERP_DB_NAME")

sandhya_erp_db_url = (
    f"mysql+pymysql://{DB_USERNAME}:{DB_PASSWORD}@{DB_HOST}:{DB_PORT}/{DB_NAME}"
)

In [4]:
import pandas as pd
import sqlalchemy

# show all rows and columns
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)

In [5]:
engine = sqlalchemy.create_engine(sandhya_erp_db_url)

In [6]:
supplier_df = pd.read_sql("SELECT * FROM erpx_dev.suppliers", engine)
supplier_df.head()

supplier_id code                      name supplier_category gst_number  \
0            1   A1               Karuna Raju             Agent       None   
1            2   A2    Sri Lakshmi Aqua Farms             Agent       None   
2            3   A3  Sri Venkateswara Traders             Agent       None   
3            4   A4      Varalakshmi Agencies             Agent       None   
4            7   A5    Sri Lakshmi Aqua Farms             Agent       None   

   broker_id contact_person street apartment  city state district postal_code  \
0          0      Gopi Raju   None      None  None    ap   Ongole      123456   
1          0        K Vijay   None      None  None    ap   bhVEM       123456   
2          0            SVT   None      None  None    ap   bhVEM       123456   
3          0         Rakesh   None      None  None    ap       KK      123456   
4          0         Rakesh   None      None  None    AP       KK      123456   

   commission bank_acc_no bank_ifsc bank_name bank_branch contact_number  \
0         2.0        None      None      None        None           None   
1         2.0        None      None      None        None           None   
2         2.0        None      None      None        None           None   
3         1.0        None      None      None        None           None   
4         1.0        None      None      None        None           None   

  email_id  is_active                 created_at created_user  \
0     None          1 2023-01-20 17:18:17.358860         None   
1     None          1 2023-01-21 15:59:41.769207         None   
2     None          1 2023-01-21 16:19:28.500526         None   
3     None          1 2023-01-23 14:39:43.386196         None   
4     None          0 2023-01-23 17:33:11.528853         None   

                  updated_at updated_user  version_flag payment_period  
0 2023-01-23 15:24:21.920682         None             1           None  
1 2023-01-23 15:24:27.197321         None             1           None  
2 2023-01-23 15:24:30.204286         None             1           None  
3 2023-01-23 15:24:34.793301         None             1           None  
4 2023-01-24 19:23:35.251509         None             2           None

In [7]:
supplier_df.describe().T

count                           mean  \
supplier_id   572.0                     288.486014   
broker_id     572.0                       0.215035   
commission    567.0                     920.155203   
is_active     572.0                       0.977273   
created_at      572  2023-08-27 04:53:19.229318400   
updated_at      572  2023-09-01 05:44:39.985777152   
version_flag  572.0                       1.083916   

                                     min                            25%  \
supplier_id                          1.0                         145.75   
broker_id                            0.0                            0.0   
commission                           1.0                            1.0   
is_active                            0.0                            1.0   
created_at    2023-01-20 17:18:17.358860  2023-02-25 11:28:33.746909440   
updated_at    2023-01-23 15:24:21.920682  2023-02-25 15:49:37.842742528   
version_flag                         1.0                            1.0   

                                        50%                            75%  \
supplier_id                           288.5                         431.25   
broker_id                               0.0                            0.0   
commission                              1.0                            1.0   
is_active                               1.0                            1.0   
created_at    2023-06-01 14:03:45.611426048  2023-11-17 12:35:04.927335936   
updated_at    2023-06-03 04:25:53.728720384  2023-11-28 12:55:13.254458880   
version_flag                            1.0                            1.0   

                                     max           std  
supplier_id                        574.0     165.29061  
broker_id                          123.0      5.142888  
commission                      521157.0  21886.494395  
is_active                            1.0      0.149163  
created_at    2025-08-12 12:22:13.410409           NaN  
updated_at    2025-08-12 12:22:13.410409           NaN  
version_flag                         4.0      0.307444

No significant numerical data in erpx_dev.suppliers table

In [8]:
# supplier with max commission

supplier_df[supplier_df["commission"] == supplier_df["commission"].max()]

supplier_id  code             name supplier_category gst_number  \
489          492  A481  SUMUKHA TRADERS             Agent       None   

     broker_id contact_person street apartment  city state district  \
489          0         PARASA   None      None  None    AP  KRISHNA   

    postal_code  commission bank_acc_no bank_ifsc bank_name bank_branch  \
489        None    521157.0        None      None      None        None   

    contact_number email_id  is_active                 created_at  \
489           None     None          1 2024-07-08 17:07:10.619320   

    created_user                 updated_at updated_user  version_flag  \
489         None 2024-07-08 17:07:10.619320         None             1   

    payment_period  
489           None

In [9]:
supplier_df.nunique()

supplier_id          572
code                 572
name                 556
supplier_category      3
gst_number             1
broker_id              2
contact_person        11
street                 1
apartment              0
city                   1
state                  5
district               7
postal_code            7
commission             3
bank_acc_no            0
bank_ifsc              0
bank_name              0
bank_branch            0
contact_number         1
email_id               0
is_active              2
created_at           572
created_user           0
updated_at           572
updated_user           0
version_flag           4
payment_period         0
dtype: int64

In [10]:
supplier_df["supplier_category"].value_counts()

supplier_category
Agent     561
Farmer     10
Broker      1
Name: count, dtype: int64

Among 572 suppliers, 561 are agents, which is almost all of the suppliers. 

In [11]:
supplier_df[supplier_df["supplier_category"] == "Broker"]

supplier_id code   name supplier_category gst_number  broker_id  \
120          123   B1  asdff            Broker       None          0   

    contact_person street apartment  city state district postal_code  \
120           None   None      None  None  None     None        None   

     commission bank_acc_no bank_ifsc bank_name bank_branch contact_number  \
120         NaN        None      None      None        None           None   

    email_id  is_active                 created_at created_user  \
120     None          1 2023-02-21 16:24:21.113710         None   

                    updated_at updated_user  version_flag payment_period  
120 2023-02-21 16:24:21.113710         None             1           None

In [12]:
supplier_df[supplier_df["supplier_category"] == "Farmer"]

supplier_id  code                                           name  \
5              8    F1                                         K Sriu   
6              9    F2                                 srinivasarao k   
21            24    F3                                          rahul   
83            86    F4                                         madhav   
234          237  A230             SANDHYA AQUA FARMS KRITHIVENNU OWN   
359          362    F5                                    V DURGA RAO   
414          417    F6  SANDHYA AQUA EXPORTS PVT LTD FARM BEERAMGUNTA   
437          440    F7                         SANDHYA AQUA SRI FARMS   
487          490    F8                               GUDURU OWN PONDS   
561          564    F9                                              x   

    supplier_category   gst_number  broker_id contact_person street apartment  \
5              Farmer         None          0         Rakesh   None      None   
6              Farmer         None          0        sravani   None      None   
21             Farmer  GST33654789          0           None   None      None   
83             Farmer         None        123           None   None      None   
234            Farmer         None          0           None   None      None   
359            Farmer         None          0        SRAVANI   None      None   
414            Farmer         None          0        KAVITHA   None      None   
437            Farmer         None          0        KAVITHA   None      None   
487            Farmer         None          0         PARASA   None      None   
561            Farmer         None          0           None   None      None   

     city state  district postal_code  commission bank_acc_no bank_ifsc  \
5    None    AP        KK      123456         1.0        None      None   
6    None    AP   Krishna      521157         NaN        None      None   
21   None  None      None        None         NaN        None      None   
83   None  None      None        None         NaN        None      None   
234  None  None      None        None         1.0        None      None   
359  None  None   KRISHNA      521157         1.0        None      None   
414  None    AP  KRISHANA      521157         1.0        None      None   
437  None    AP   KRISHNA      521157         1.0        None      None   
487  None    AP   KRISHNA      521157         1.0        None      None   
561  None  None      None        None         NaN        None      None   

    bank_name bank_branch contact_number email_id  is_active  \
5        None        None           None     None          1   
6        None        None           None     None          1   
21       None        None           None     None          1   
83       None        None           None     None          1   
234      None        None           None     None          1   
359      None        None           None     None          1   
414      None        None           None     None          1   
437      None        None           None     None          1   
487      None        None           None     None          1   
561      None        None           None     None          1   

                    created_at created_user                 updated_at  \
5   2023-01-24 11:08:26.341499         None 2023-01-24 11:08:26.341499   
6   2023-01-24 11:36:09.012417         None 2023-01-24 19:29:02.666640   
21  2023-01-24 19:38:08.526412         None 2023-01-24 19:38:08.526412   
83  2023-02-08 18:23:33.541614         None 2023-02-21 16:24:21.000000   
234 2023-04-21 10:41:13.773245         None 2023-08-29 17:01:45.000000   
359 2023-08-01 10:20:03.658414         None 2023-08-01 10:20:03.658414   
414 2023-10-21 14:12:00.607933         None 2023-10-21 14:12:00.607933   
437 2023-12-12 11:09:54.815987         None 2023-12-12 11:09:54.815987   
487 2024-06-27 12:43:06.874291         None 2024-06-27 12:43:06.874291   
561 2025-06-18 13:06:43.879680        

In [13]:
indent_df = pd.read_sql("SELECT * FROM erpx_dev_rm_procurement.indent", engine)
indent_df.head()

is_active                 created_at created_user          updated_at  \
0          1 2023-01-20 18:36:47.331519      unitone 2023-01-23 15:43:48   
1          1 2023-01-20 18:48:41.656502      unitone 2023-01-20 18:49:11   
2          1 2023-01-21 16:02:02.897476   superadmin 2023-01-21 16:09:08   
3          1 2023-01-21 16:21:02.870258   superadmin 2023-01-23 10:30:51   
4          1 2023-01-23 14:42:02.614058      unittwo 2023-01-23 14:48:31   

  updated_user  version_flag  indent_id   indent_code         indent_date  \
0      unitone             6          3  I24/A20/0001 2023-01-20 13:05:15   
1      unitone             3          4  I24/A20/0002 2023-01-20 13:18:12   
2   superadmin             5          5  I24/A21/0003 2023-01-21 10:29:58   
3   superadmin             5          6  I24/A21/0004 2023-01-21 10:49:33   
4      unittwo             6          7  I24/A23/0005 2023-01-23 09:09:50   

  supplier_type commission_type  commission     harvesting_date  \
0         Agent      Percentage           0 2023-01-21 13:05:15   
1         Agent      Percentage           0 2023-01-21 13:18:12   
2         Agent      Percentage           0 2023-01-22 10:29:58   
3         Agent      Percentage           0 2023-01-22 10:49:33   
4         Agent      Percentage           0 2023-01-24 09:09:50   

  harvesting_time  assign_lead_1  assign_lead_2  vehicle_assigned  \
0         Morning           1809            NaN                 0   
1         Morning           1809            NaN                 0   
2         Morning           1809            NaN                 0   
3         Morning           1809            NaN                 0   
4         Morning           1809            NaN                 0   

   grn_completed  plant_id  indent_by  is_vehicle_required  harvest_type  
0              1         1       2178                    0  Re Weighment  
1              1         2       1810                    0  Re Weighment  
2              1         2       2178                    0  Live Harvest  
3              1         1       2178                    0  Re Weighment  
4              1         3       1809                    0  Live Harvest

In [14]:
indent_items_df = pd.read_sql(
    "SELECT * FROM erpx_dev_rm_procurement.indent_items", engine
)
indent_items_df.head()

indent_item_id  broker  farmer  farmer_code  product  expected_count  \
0               2       0       1          NaN        2              97   
1               3       0       1          NaN        2              45   
2               4       0       2          NaN        2              25   
3               5       0       3          NaN        2              55   
4               6       0       4          NaN        2              70   

   expected_qty  expected_price  expected_count2  expected_price2  \
0          1500             237            100.0            225.0   
1          1000             350              NaN              NaN   
2          5000             300              NaN              NaN   
3          3500             275             57.0            270.0   
4          2500             290             75.0            286.0   

  expected_count3 expected_price3  is_active                 created_at  \
0            None            None          1 2023-01-20 18:36:47.337082   
1            None            None          1 2023-01-20 18:48:41.662178   
2            None            None          1 2023-01-21 16:02:02.904148   
3            None            None          1 2023-01-21 16:21:02.877064   
4            None            None          1 2023-01-23 14:42:02.627996   

  created_user          updated_at updated_user  version_flag  indent_id  \
0        Admin 2023-01-20 18:37:04         None             2          3   
1        Admin 2023-01-20 18:48:52         None             2          4   
2        Admin 2023-01-21 16:02:11         None             2          5   
3        Admin 2023-01-21 16:21:33         None             2          6   
4        Admin 2023-01-23 14:42:18         None             2          7   

   harvest_report_id  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0

In [15]:
indent_items_df.describe().T

count                           mean  \
indent_item_id     14677.0                         7340.0   
broker             14677.0                            0.0   
farmer             14677.0                     179.438373   
farmer_code            1.0                            1.0   
product            14677.0                       1.951898   
expected_count     14677.0                       57.18546   
expected_qty       14677.0                    2840.928664   
expected_price     14677.0                      316.85835   
expected_count2       12.0                     152.083333   
expected_price2        8.0                        337.375   
is_active          14677.0                            1.0   
created_at           14677  2023-11-25 09:42:20.296373504   
updated_at           14677  2023-11-26 07:15:25.184625408   
version_flag       14677.0                       1.545139   
indent_id          14677.0                     4124.35048   
harvest_report_id  14677.0                            0.0   

                                          min                            25%  \
indent_item_id                            2.0                         3671.0   
broker                                    0.0                            0.0   
farmer                                    1.0                           40.0   
farmer_code                               1.0                            1.0   
product                                   1.0                            2.0   
expected_count                            1.0                           35.0   
expected_qty                             10.0                         1100.0   
expected_price                            0.0                          265.0   
expected_count2                          34.0                          52.75   
expected_price2                          90.0                         258.75   
is_active                                 1.0                            1.0   
created_at         2023-01-20 18:36:47.337082  2023-05-12 12:00:16.910627072   
updated_at                2023-01-20 18:37:04            2023-05-12 16:03:01   
version_flag                              1.0                            1.0   
indent_id                                 3.0                         2072.0   
harvest_report_id                         0.0                            0.0   

                                             50%  \
indent_item_id                            7340.0   
broker                                       0.0   
farmer                                     133.0   
farmer_code                                  1.0   
product                                      2.0   
expected_count                              55.0   
expected_qty                              2000.0   
expected_price                             300.0   
expected_count2                            105.0   
expected_price2                            274.0   
is_active                                    1.0   
created_at         2023-08-29 16:18:27.012815104   
updated_at                   2023-08-29 17:43:43   
version_flag                                 1.0   
indent_id                                 4227.0   
harvest_report_id                            0.0   

                                             75%                         max  \
indent_item_id                           11009.0                     14678.0   
broker                                       0.0                         0.0   
farmer                                     284.0                       574.0   
farmer_code                                  1.0                         1.0   
product                                      2.0                         3.0   
expected_count                              75.0                      1800.0   
expected_qty                              3700.0                     36000.0   
expected_price                             350.0                      8000.0   
expected_count2  

In [16]:
indent_items_df.nunique()

indent_item_id       14677
broker                   1
farmer                 551
farmer_code              1
product                  3
expected_count         115
expected_qty           187
expected_price         173
expected_count2         11
expected_price2          7
expected_count3          0
expected_price3          0
is_active                1
created_at           14677
created_user             1
updated_at           11651
updated_user             0
version_flag            10
indent_id             8221
harvest_report_id        1
dtype: int64

In [17]:
grn_df = pd.read_sql("SELECT * FROM erpx_dev_rm_procurement.grn", engine)
grn_df.head()

grn_id            grn_date         grn_number supplier_type  is_active  \
0      10 2023-01-20 18:42:08  G24/KRN/null/0001         Agent          1   
1      11 2023-01-20 18:48:55  G24/KRN/null/0002         Agent          1   
2      12 2023-01-20 18:54:14  G24/KRN/null/0003         Agent          1   
3      13 2023-01-21 16:02:15  G24/123/null/0004         Agent          1   
4      14 2023-01-21 16:21:36  G24/SVT/null/0005         Agent          1   

                  created_at created_user                 updated_at  \
0 2023-01-20 18:43:15.625959      unitone 2023-01-23 12:11:13.000000   
1 2023-01-20 18:49:11.632700      unitone 2023-03-30 14:26:57.151334   
2 2023-01-20 18:54:36.696090      unitone 2023-01-23 15:43:48.000000   
3 2023-01-21 16:06:28.127043   superadmin 2023-01-21 16:09:34.000000   
4 2023-01-21 16:23:02.922709   superadmin 2023-01-23 10:30:51.000000   

  updated_user  version_flag  indent_no  plant_id  sub_plant_id  
0      unitone             4          3         2           6.0  
1         None             1          4         0           NaN  
2        admin             2          3         1           NaN  
3   superadmin             4          5         2           4.0  
4        admin             4          6         1           8.0

In [18]:
grn_items_df = pd.read_sql("SELECT * FROM erpx_dev_rm_procurement.grn_items", engine)
grn_items_df.head()

grn_item_id  grn_id  indent_id         lot_number  sale_order_id  \
0            1      10          3   LOT-U1/24/1/0001            NaN   
1            2      11          4   LOT-U1/24/1/0049            NaN   
2            3      12          3               None            NaN   
3            4      13          5   LOT-U1/24/2/0416            NaN   
4            5      14          6  LOT-KKD/24/1/0002            NaN   

   sale_order_item_id  farmer  farmer_code  quantity  count    price  \
0                 NaN       1          NaN      1000     55  50000.0   
1                 NaN       1          NaN      1000     50      NaN   
2                 NaN       1          NaN      1490     97   5200.0   
3                 NaN       2          NaN      2606     31    290.0   
4                 NaN       3          NaN      3264     57    270.0   

   received_boxes  soft_percentage  boxes  is_active  assign_to_rm  \
0           120.0              2.0    120          1             1   
1             0.0              NaN    150          1             1   
2            48.0              0.0     48          1             0   
3            87.0              0.0     87          1             1   
4           116.0              0.0    116          1             1   

                  created_at created_user          updated_at updated_user  \
0 2023-01-20 18:43:15.631362      unitone 2023-01-23 12:11:18      unitone   
1 2023-01-20 18:49:11.637446      unitone 2023-01-25 12:04:13         None   
2 2023-01-20 18:54:36.702904      unitone 2023-01-23 15:43:48        admin   
3 2023-01-21 16:06:28.135154   superadmin 2023-02-08 15:42:25   superadmin   
4 2023-01-21 16:23:02.927747   superadmin 2023-01-23 12:24:27        admin   

   version_flag  indent_item_id  product  broker  vi_inspection  \
0             3               2        2       0              0   
1             2               3        2       0              0   
2             2               2        2       0              0   
3             4               4        4       0              0   
4             4               5        5       0              0   

   rm_inspection  antibiotic_test actual_quantity actual_count loose_shell  \
0              0                0            None         None           3   
1              0                0            None         None        None   
2              0                0            None         None           0   
3              0                0            None         None           0   
4              0                0            None         None           0   

   fungus_percentage  ice_boxes  empty_boxes  rm_boxes seal_no remarks  \
0                1.0        NaN          NaN       NaN    None    None   
1                NaN        NaN          NaN       NaN    None    None   
2                0.0        NaN          NaN       NaN    None    None   
3                0.0        NaN          NaN       NaN    None    None   
4                0.0        NaN          NaN       NaN    None    None   

  plant_lot_number  ref_no  
0             None     NaN  
1             None     NaN  
2             None     NaN  
3             None     NaN  
4             None     NaN

In [19]:
grn_items_df.describe().T

count                           mean  \
grn_item_id         14602.0                    7446.208328   
grn_id              14602.0                    4279.944871   
indent_id           14602.0                       4122.598   
sale_order_id       12608.0                    2974.662754   
sale_order_item_id  12608.0                   10952.204711   
farmer              14602.0                     179.319203   
farmer_code             1.0                            1.0   
quantity            14602.0                     2823.37714   
count               14602.0                      60.077455   
price                 269.0                     517.925651   
received_boxes      13820.0                      94.820478   
soft_percentage     12113.0                       0.026253   
boxes               14602.0                      93.206616   
is_active           14602.0                            1.0   
assign_to_rm        14602.0                        0.99548   
created_at            14602  2023-11-26 10:46:11.862279680   
updated_at            14602  2025-02-15 20:42:39.818146816   
version_flag        14602.0                       4.085468   
indent_item_id      14602.0                    7304.658471   
product             14602.0                    6818.045268   
broker              14602.0                            0.0   
vi_inspection       14602.0                       0.000205   
rm_inspection       14602.0                            0.0   
antibiotic_test     14602.0                       0.000068   
fungus_percentage   12113.0                       0.011062   
ice_boxes              66.0                       5.560606   
empty_boxes            66.0                      19.939394   
rm_boxes               66.0                           98.5   
ref_no              12565.0                    1258.219897   

                                           min                            25%  \
grn_item_id                                1.0                        3809.25   
grn_id                                    10.0                         2250.0   
indent_id                                  3.0                         2065.0   
sale_order_id                            220.0                         1927.0   
sale_order_item_id                       358.0                         8397.0   
farmer                                     1.0                           40.0   
farmer_code                                1.0                            1.0   
quantity                                   4.0                         1150.0   
count                                      5.0                           40.0   
price                                    103.0                          278.0   
received_boxes                             0.0                           40.0   
soft_percentage                            0.0                            0.0   
boxes                                      1.0                           39.0   
is_active                                  1.0                            1.0   
assign_to_rm                               0.0                            1.0   
created_at          2023-01-20 18:43:15.631362  2023-05-12 15:50:10.506600192   
updated_at                 2023-01-23 12:11:18  2025-05-09 09:22:22.411624960   
version_flag                               1.0                            3.0   
indent_item_id                             1.0                        3654.25   
product                                    1.0                        2987.25   
broker                                     0.0                            0.0   
vi_inspection                              0.0                            0.0   
rm_inspection                              0.0                            0.0   
antibiotic_test                            0.0                            0.0   
fungus_percentage                          0.0                            0.0   
ice_boxes                                  0.0         

In [20]:
grn_items_df["plant_lot_number"].nunique()

12997

In [21]:
grn_items_df.nunique()

grn_item_id           14602
grn_id                 8190
indent_id              8186
lot_number            13504
sale_order_id           696
sale_order_item_id      804
farmer                  551
farmer_code               1
quantity               5976
count                   175
price                   101
received_boxes          484
soft_percentage          12
boxes                   483
is_active                 1
assign_to_rm              2
created_at            14602
created_user             23
updated_at             2383
updated_user             21
version_flag             17
indent_item_id        14112
product               12110
broker                    1
vi_inspection             2
rm_inspection             1
antibiotic_test           2
actual_quantity           0
actual_count              0
loose_shell              20
fungus_percentage         6
ice_boxes                15
empty_boxes              25
rm_boxes                 52
seal_no                  38
remarks             

In [22]:
query = """
    SELECT

    -- Farmer Information
    ii.farmer,
    s.name AS farmer_name,

    -- Time to GRN
    AVG(TIMESTAMPDIFF(HOUR, i.indent_date, g.grn_date)) AS avg_time_to_grn,

    -- Average Expected (Planned) Values from Indent
    SUM(ii.expected_qty) AS total_expected_qty,                                 -- SUMMING ALL EXPECTED QUANITTY FROM FARMER RATHER TO AVERAGING
    AVG(ii.expected_price) AS avg_expected_price,
    -- AVG(ii.expected_count) AS avg_expected_count, 

    -- Average Actual Received Values from GRN
    SUM(gi.quantity) AS total_actual_quantity,                           -- SUMMING ALL ACTUAL QUANITTY FROM FARMER RATHER TO AVERAGING   
    AVG(gi.price) AS avg_actual_price,
    -- AVG(gi.actual_count) AS avg_actual_count,
    AVG(gi.received_boxes) AS avg_received_boxes,

    -- Key Quality & Performance Metrics
    AVG(ABS(gi.quantity - ii.expected_qty)) AS avg_quantity_variance,
    AVG(gi.price - ii.expected_price) AS avg_price_variance,
    AVG(ABS(gi.count - ii.expected_count)) AS avg_expected_count_variance,
    

    -- Identifiers for Traceability
    GROUP_CONCAT(DISTINCT i.indent_id) AS indent_ids,
    GROUP_CONCAT(DISTINCT ii.indent_item_id) AS indent_item_ids,
    GROUP_CONCAT(DISTINCT g.grn_id) AS grn_ids,
    GROUP_CONCAT(DISTINCT gi.grn_item_id) AS grn_item_ids,
    
    COUNT(DISTINCT i.indent_id) AS indent_count,
    COUNT(DISTINCT g.grn_id) AS grn_count,
    COUNT(DISTINCT gi.indent_item_id) AS indent_item_count,
    COUNT(DISTINCT gi.grn_item_id) AS grn_item_count,

    COUNT(DISTINCT gi.plant_lot_number) AS lots_supplied

FROM
    erpx_dev_rm_procurement.indent_items AS ii
JOIN
    erpx_dev_rm_procurement.indent AS i ON ii.indent_id = i.indent_id
LEFT JOIN
    erpx_dev_rm_procurement.grn AS g ON g.indent_no = i.indent_id
LEFT JOIN 
    erpx_dev_rm_procurement.grn_items AS gi ON gi.grn_id = g.grn_id
LEFT JOIN
    erpx_dev.suppliers AS s ON s.supplier_id = ii.farmer
GROUP BY
    ii.farmer
ORDER BY
    ii.farmer;
"""

test_df = pd.read_sql(query, engine)
test_df.head()

farmer               farmer_name  avg_time_to_grn  total_expected_qty  \
0       1               Karuna Raju         302.8000             21347.0   
1       2    Sri Lakshmi Aqua Farms          33.7000             43700.0   
2       3  Sri Venkateswara Traders           5.0000              3500.0   
3       4      Varalakshmi Agencies         128.7684            367500.0   
4       8                    K Sriu           3.3333             10500.0   

   avg_expected_price  total_actual_quantity  avg_actual_price  \
0            301.4000                23530.0        27600.0000   
1            293.8000                35934.0          278.0000   
2            275.0000                 3264.0          270.0000   
3            319.5526               336530.0          312.5839   
4            286.6667                10072.0          287.5000   

   avg_received_boxes  avg_quantity_variance  avg_price_variance  \
0             84.5556               421.3000          27363.0000   
1            122.0000              1389.8000             -3.6000   
2            116.0000               236.0000             -5.0000   
3             69.5122               852.7263              5.7591   
4             60.8000               343.6667              7.5000   

   avg_expected_count_variance  \
0                       7.3000   
1                       4.5000   
2                       2.0000   
3                       9.6895   
4                       5.3333   

                                          indent_ids  \
0                    3,4,9,56,57,1399,2671,4637,7297   
1                                     5,10,1603,7529   
2                                                  6   
3  7,8,15,17,51,53,58,84,97,113,128,141,154,181,2...   
4                                       11,7572,7911   

                                     indent_item_ids  \
0                  2,3,10,98,99,2459,4772,8274,13100   
1                            4,11,12,2832,2833,13486   
2                                                  5   
3  6,7,8,9,20,21,22,23,24,26,27,88,89,90,92,93,94...   
4                                  13,14,13561,14169   

                                             grn_ids  \
0          10,11,12,17,1119,1243,1467,2841,4786,7447   
1                                    13,18,1788,7679   
2                                                 14   
3  15,16,23,25,78,80,83,113,127,143,158,170,184,2...   
4                                       19,7722,8063   

                                        grn_item_ids  indent_count  grn_count  \
0            1,2,3,10,1821,2025,2420,4908,8345,13178             9         10   
1                            4,11,12,2987,2988,13564             4          4   
2                                                  5             1          1   
3  6,7,8,9,20,21,22,23,24,26,27,74,75,76,77,79,80...            16         16   
4                                  13,14,13638,14251             3          3   

   indent_item_count  grn_item_count  lots_supplied  
0                  9              10              5  
1                  6               6              3  
2                  1               1              0  
3                 52              52              2  
4                  4               4              2

In [23]:
test_df.describe().T

count           mean            std       min  \
farmer                       551.0     291.119782     165.036096    1.0000   
avg_time_to_grn              551.0      99.119519     378.688365    3.3333   
total_expected_qty           551.0  143762.956443  402250.443328  140.0000   
avg_expected_price           551.0     329.588312      77.066322    0.0000   
total_actual_quantity        551.0  142122.050817  401283.414715  118.0000   
avg_actual_price              61.0     761.697493    3493.980183  205.0000   
avg_received_boxes           532.0      89.399343      62.991013    3.0000   
avg_quantity_variance        551.0     540.135413     482.368051    0.0000   
avg_price_variance            61.0     455.398266    3502.654997  -39.0000   
avg_expected_count_variance  551.0      10.186916      36.740303    0.0000   
indent_count                 551.0      14.978221      41.329051    1.0000   
grn_count                    551.0      14.921960      41.230475    1.0000   
indent_item_count            551.0      26.125227      66.732606    1.0000   
grn_item_count               551.0      26.653358      68.097908    1.0000   
lots_supplied                551.0      23.760436      62.010708    0.0000   

                                     25%         50%           75%  \
farmer                        152.500000    294.0000    433.000000   
avg_time_to_grn                54.365150     70.8868     91.973250   
total_expected_qty           7100.000000  23000.0000  86240.000000   
avg_expected_price            284.475000    314.0000    359.056600   
total_actual_quantity        6892.000000  22867.0000  79593.000000   
avg_actual_price              285.000000    305.3333    342.000000   
avg_received_boxes             52.865475     75.0000    104.216825   
avg_quantity_variance         191.833300    427.7500    766.866350   
avg_price_variance             -1.000000      5.0000     10.000000   
avg_expected_count_variance     5.000000      7.1552     10.921850   
indent_count                    1.000000      3.0000      9.000000   
grn_count                       1.000000      3.0000      9.000000   
indent_item_count               2.000000      5.0000     16.500000   
grn_item_count                  2.000000      5.0000     16.500000   
lots_supplied                   1.500000      4.0000     15.000000   

                                      max  
farmer                       5.740000e+02  
avg_time_to_grn              8.832000e+03  
total_expected_qty           3.418625e+06  
avg_expected_price           9.000000e+02  
total_actual_quantity        3.475924e+06  
avg_actual_price             2.760000e+04  
avg_received_boxes           4.920000e+02  
avg_quantity_variance        3.708111e+03  
avg_price_variance           2.736300e+04  
avg_expected_count_variance  8.575000e+02  
indent_count                 4.730000e+02  
grn_count                    4.730000e+02  
indent_item_count            5.820000e+02  
grn_item_count               5.920000e+02  
lots_supplied                5.430000e+02

In [24]:
# Drop avg_actual_price and avg_price_variance due to low volume of data
dataset = test_df.drop(
    columns=[
        "avg_actual_price",
        "avg_price_variance",
        "grn_ids",
        "grn_item_ids",
        "indent_item_ids",
        "indent_ids",
        "lots_supplied",
    ]
)

In [25]:
dataset.describe().T

count           mean            std       min  \
farmer                       551.0     291.119782     165.036096    1.0000   
avg_time_to_grn              551.0      99.119519     378.688365    3.3333   
total_expected_qty           551.0  143762.956443  402250.443328  140.0000   
avg_expected_price           551.0     329.588312      77.066322    0.0000   
total_actual_quantity        551.0  142122.050817  401283.414715  118.0000   
avg_received_boxes           532.0      89.399343      62.991013    3.0000   
avg_quantity_variance        551.0     540.135413     482.368051    0.0000   
avg_expected_count_variance  551.0      10.186916      36.740303    0.0000   
indent_count                 551.0      14.978221      41.329051    1.0000   
grn_count                    551.0      14.921960      41.230475    1.0000   
indent_item_count            551.0      26.125227      66.732606    1.0000   
grn_item_count               551.0      26.653358      68.097908    1.0000   

                                     25%         50%           75%  \
farmer                        152.500000    294.0000    433.000000   
avg_time_to_grn                54.365150     70.8868     91.973250   
total_expected_qty           7100.000000  23000.0000  86240.000000   
avg_expected_price            284.475000    314.0000    359.056600   
total_actual_quantity        6892.000000  22867.0000  79593.000000   
avg_received_boxes             52.865475     75.0000    104.216825   
avg_quantity_variance         191.833300    427.7500    766.866350   
avg_expected_count_variance     5.000000      7.1552     10.921850   
indent_count                    1.000000      3.0000      9.000000   
grn_count                       1.000000      3.0000      9.000000   
indent_item_count               2.000000      5.0000     16.500000   
grn_item_count                  2.000000      5.0000     16.500000   

                                      max  
farmer                       5.740000e+02  
avg_time_to_grn              8.832000e+03  
total_expected_qty           3.418625e+06  
avg_expected_price           9.000000e+02  
total_actual_quantity        3.475924e+06  
avg_received_boxes           4.920000e+02  
avg_quantity_variance        3.708111e+03  
avg_expected_count_variance  8.575000e+02  
indent_count                 4.730000e+02  
grn_count                    4.730000e+02  
indent_item_count            5.820000e+02  
grn_item_count               5.920000e+02

Performing TOPSIS method to rank the farmer performance
1. Normalize the data
2. Calculate the weighted normalized decision matrix
3. Calculate the separation measures
4. Calculate the relative closeness to the ideal solution
5. Rank the farmers based on the relative closeness to the ideal solution



In [26]:
# Normalizing the data and normalizing if we have weights
def normalize_data(dataset, nCol, weights=None):
    for i in range(1, nCol):
        temp = 0

        for j in range(len(dataset)):
            temp += dataset.iloc[j, i] ** 2
        temp = temp**0.5

        for j in range(len(dataset)):
            if weights is not None:
                dataset.iloc[j, i] = (dataset.iloc[j, i] / temp) * weights[i - 1]
            else:
                dataset.iloc[j, i] = dataset.iloc[j, i] / temp

    return dataset


def calculate_ideal_values(dataset, nCol, impact):
    ideal_best = []
    ideal_worst = []

    for i in range(1, nCol):
        if impact[i - 1] == "+":
            ideal_best.append(dataset.iloc[:, i].max())
            ideal_worst.append(dataset.iloc[:, i].min())
        else:
            ideal_best.append(dataset.iloc[:, i].min())
            ideal_worst.append(dataset.iloc[:, i].max())

    return ideal_best, ideal_worst


def calculate_separation_measures(dataset, ideal_best, ideal_worst, nCol):
    separation_best = []
    separation_worst = []

    for i in range(len(dataset)):
        best = 0
        worst = 0

        for j in range(1, nCol):
            best += (dataset.iloc[i, j] - ideal_best[j - 1]) ** 2
            worst += (dataset.iloc[i, j] - ideal_worst[j - 1]) ** 2

        separation_best.append(best**0.5)
        separation_worst.append(worst**0.5)

    return separation_best, separation_worst


def calculate_relative_closeness(separation_best, separation_worst):
    relative_closeness = []

    for i in range(len(separation_best)):
        if separation_best[i] + separation_worst[i] == 0:
            relative_closeness.append(0)
        else:
            relative_closeness.append(
                separation_worst[i] / (separation_best[i] + separation_worst[i])
            )

    return relative_closeness


def rank_farmers(dataset, nCol, impact, weights=None):
    normalized_data = normalize_data(dataset.copy(), nCol, weights)
    ideal_best, ideal_worst = calculate_ideal_values(normalized_data, nCol, impact)
    separation_best, separation_worst = calculate_separation_measures(
        normalized_data, ideal_best, ideal_worst, nCol
    )
    relative_closeness = calculate_relative_closeness(separation_best, separation_worst)
    normalized_data["relative_closeness"] = relative_closeness
    normalized_data["rank"] = normalized_data["relative_closeness"].rank(
        ascending=False
    )

    return normalized_data.sort_values(by="rank")


numeric_columns_impact_mapper = {
    "avg_time_to_grn": "-",
    "total_expected_qty": "+",
    "avg_expected_price": "-",
    "total_actual_quantity": "+",
    "avg_received_boxes": "+",
    "avg_quantity_variance": "-",
    "avg_expected_count_variance": "-",
    "indent_count": "+",
    "grn_count": "+",
    "indent_item_count": "+",
    "grn_item_count": "+",
}

farmer_ids = dataset[["farmer", "farmer_name"]].copy()
numeric_dataset = dataset[
    ["farmer"] + list(numeric_columns_impact_mapper.keys())
].copy()

numeric_dataset = numeric_dataset.fillna(numeric_dataset.mean())

nCol = numeric_dataset.shape[1]

impact = list(numeric_columns_impact_mapper.values())  # Adjusted for 10 numeric columns

equal_weight_factor = 1 / (nCol - 1)  # Equal weight factor for normalization

weights = [equal_weight_factor] * (nCol - 1)

ranked_farmers = rank_farmers(numeric_dataset, nCol, impact, weights)

# Merge with original dataset to get all columns, then add ranking information
final_results = dataset.merge(
    ranked_farmers[["farmer", "relative_closeness", "rank"]], on="farmer", how="left"
)
final_results = final_results.sort_values(by="rank")
final_results.head()

/tmp/ipykernel_165201/2571334127.py:12: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.0007935419938935141' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  dataset.iloc[j, i] = (dataset.iloc[j, i] / temp) * weights[i - 1]
/tmp/ipykernel_165201/2571334127.py:12: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.0008839625463257504' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  dataset.iloc[j, i] = (dataset.iloc[j, i] / temp) * weights[i - 1]
/tmp/ipykernel_165201/2571334127.py:12: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.0004867582402885267' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  dataset.iloc[j, i] = (dat

farmer     farmer_name  avg_time_to_grn  total_expected_qty  \
17       21         KS RAJU          74.3528           3391850.0   
154     169   RAMALINGARAJU         200.7969           3106274.0   
15       19  SRINIVASARAO K          77.4832           3418625.0   
14       18         K SRINU         107.8041           2036650.0   
36       40  SRI RAMA AQUA           81.9210           2543562.0   

     avg_expected_price  total_actual_quantity  avg_received_boxes  \
17             289.8090              3463088.0            133.9238   
154            306.6151              3125085.0             97.1199   
15             312.4812              3475924.0             94.8579   
14             309.8884              1922421.0             77.6538   
36             310.9212              2520776.0             70.5551   

     avg_quantity_variance  avg_expected_count_variance  indent_count  \
17               1169.1873                      11.3017           473   
154               648.4749                       8.5735           400   
15               1732.8690                      17.1245           361   
14               1399.0011                      18.2130           239   
36                678.7182                      10.8832           179   

     grn_count  indent_item_count  grn_item_count  relative_closeness  rank  
17         473                563             576            0.910558   1.0  
154        400                556             556            0.887877   2.0  
15         360                582             592            0.866148   3.0  
14         239                386             406            0.761118   4.0  
36         178                342             344            0.743128   5.0

In [27]:
final_results.head(50)

farmer                         farmer_name  avg_time_to_grn  \
17       21                             KS RAJU          74.3528   
154     169                       RAMALINGARAJU         200.7969   
15       19                      SRINIVASARAO K          77.4832   
14       18                             K SRINU         107.8041   
36       40                      SRI RAMA AQUA           81.9210   
388     407     SHREEYAN MARINE EXPORTS PVT LTD          53.4301   
49       53                        NAGESH VARMA          80.6364   
265     284                          MASTAN SKM          62.6291   
35       39                  CHANDRIKA SEAFOODS          71.1003   
73       77                   PK RAMALINGA RAJU          77.2637   
66       70                        B PANDURANGA          83.1048   
99      105                     SYAMALA TRADERS          81.0442   
253     272                             K RAGHU         105.5844   
484     507                  VELANKKANI TRADERS          48.2413   
220     237  SANDHYA AQUA FARMS KRITHIVENNU OWN          63.2885   
9        13                 MEENAKSHI FISHEREIS         108.3900   
28       32                RAVINDRA ENTERPRISES          60.9786   
89       94                                SKTC          84.2936   
78       82                             SV AQUA          72.2532   
147     162                                KVSN          76.3183   
44       48                     PVR ENTERPRISES          98.4272   
31       35                          SS TRADERS          71.3807   
250     268                     ARUNODAYA AQUA           67.7689   
18       22                        SKM SEAFOODS         125.8602   
305     324                       ANJANEYA AQUA          68.6568   
45       49                            R KARUNA          90.9378   
72       76                       SIVA SEAFOODS          82.7271   
6        10                      Sita rama raju          77.9415   
390     409     SREEYAAN MARINE EXPORTS PVT LTD          58.3802   
309     328                          NVVS VARMA         120.3828   
433     454                       PV TRADELINKS          49.7907   
11       15               VENKATESWARA SEAFOODS          78.4536   
34       38                     VENKANNA BABU Y          68.2727   
256     275                                SRSF          78.7123   
155     170                      M NAVEEN REDDY          60.8792   
537     560                   LAVANYA SEA FOODS          48.8293   
126     141                      AQUA EXCHANGE           87.4361   
129     144                     P RAMALINGARAJU          85.5563   
176     191                                 SRK          68.0000   
193     208            P LAKSHMI NARAYANA VARMA          71.1786   
57       61                   NAGAMALLESWARAO K          69.1624   
207     222                        GANESH VARMA          76.6875   
201     216                     NN SRINIVASARAO         120.4777   
431     452                          NN TRADERS          54.4776   
113     127                             RAMBABU          56.4326   
401     420                ADITYA AQUA AGENCIES          57.4219   
468     490                    GUDURU OWN PONDS          56.0876   
16       20             SAI LAKSHMI AUQA FARMS           99.0094   
22       26                       SWAMY NAIDU K         110.7243   
39       43               VENKATESWARA SEAFOODS          84.1803   

     total_expected_qty  avg_expected_price  total_actual_quantity  \
17            3391850.0            289.8090              3463088.0   
154           3106274.0            306.6151              3125085.0   
15            3418625.0            312.4812              3475924.0   
14            2036650.0            309.8884              1922421.0   
36            2543562.0            310.9212              2520776.0   
388           2701530.0            304.5837              2606042.0   
49            1872480.0            298.5796   

In [28]:
# Save to Excel file
final_results.to_excel("farmer_performance_ranking.xlsx", index=False)

scoring should be moving average (1 month last) 
- time to grn
- yield in grading
- weight (qty) variance from intent to GRN
- no of specieas
- no intents
- total of actual GRN weight
- no. of lots during grading 

ranking to be updated as per above parameters. 

In [29]:
query = """
WITH hl_cte AS (
    SELECT
        gi.grn_id,
        gi.created_at AS grn_created_at,
        pg.lot_name,
        gi.count AS hon_count,
        gi.quantity AS hon_weight,
        pg.created_at AS grading_created_at,
        FLOOR(SUM(pgr.weight) - (pg.crate_weight * COUNT(pgr.crates))) AS hl_weight,
        ( (SUM(pgr.weight) - (pg.crate_weight * COUNT(pgr.crates))) / NULLIF(gi.quantity, 0) ) * 100 AS grading_yield,
        GROUP_CONCAT(DISTINCT pgr.sale_order ORDER BY pgr.sale_order ASC SEPARATOR ', ') AS grading_sale_orders
    FROM erpx_dev_rm_procurement.grn_items gi
    LEFT JOIN erpx_dev_production.pp_grading pg
        ON gi.plant_lot_number = pg.lot_name
    LEFT JOIN erpx_dev_production.pp_grading_readings pgr
        ON pg.session_id = pgr.session_id
    GROUP BY 
        gi.grn_id,
        gi.created_at,
        pg.lot_name,
        gi.count,
        gi.quantity,
        pg.created_at
)
SELECT
    ii.farmer,
    s.name AS farmer_name,

    AVG(TIMESTAMPDIFF(HOUR, i.indent_date, g.grn_date)) AS avg_time_to_grn,

    SUM(gi.quantity) AS total_actual_quantity,                                         -- SUMMING ALL ACTUAL QUANITTY FROM FARMER RATHER TO AVERAGING   

    AVG(ABS(gi.quantity - ii.expected_qty)) AS avg_quantity_variance,
    
    GROUP_CONCAT(DISTINCT i.indent_id) AS indent_ids,
    COUNT(DISTINCT i.indent_id) AS indent_count,
    
    GROUP_CONCAT(DISTINCT gi.plant_lot_number) AS plant_lot_numbers,
    COUNT(DISTINCT gi.plant_lot_number) AS lots_supplied,
    
    AVG(hl.grading_yield) AS avg_grading_yield

FROM
    erpx_dev_rm_procurement.indent_items AS ii
JOIN
    erpx_dev_rm_procurement.indent AS i ON ii.indent_id = i.indent_id
LEFT JOIN
    erpx_dev_rm_procurement.grn AS g ON g.indent_no = i.indent_id
LEFT JOIN 
    erpx_dev_rm_procurement.grn_items AS gi ON gi.grn_id = g.grn_id
LEFT JOIN
    erpx_dev.suppliers AS s ON s.supplier_id = ii.farmer
LEFT JOIN 
    hl_cte AS hl ON hl.grn_id = gi.grn_id
WHERE
    g.grn_date >= DATE_SUB(CURDATE(), INTERVAL 1 MONTH)
GROUP BY
    ii.farmer
ORDER BY
    ii.farmer;
"""

modified_rank_df = pd.read_sql(query, engine)
modified_rank_df.head()

farmer               farmer_name  avg_time_to_grn  total_actual_quantity  \
0      14            BRAHMA NAIDU K             48.0                92176.0   
1      18                   K SRINU             48.0                 5867.0   
2      21                   KS RAJU             48.0                 4741.0   
3      32      RAVINDRA ENTERPRISES             48.0                 2176.0   
4      34  SRI KANAKA DURGA TRADERS             48.0                82161.0   

   avg_quantity_variance                indent_ids  indent_count  \
0              1369.6250                      8197             1   
1               316.5000                 8094,8190             2   
2               259.0000                      8098             1   
3               324.0000                      8208             1   
4               402.1154  8125,8149,8156,8191,8202             5   

                                   plant_lot_numbers  lots_supplied  \
0                    5P2-603,5P2-604,5P2-605,5P2-606              4   
1                                    5P1-431,5P1-552              2   
2                                            5P1-440              1   
3                                            5P2-611              1   
4  5P1-466,5P1-487,5P1-488,5P1-496,5P1-497,5P1-55...              8   

   avg_grading_yield  
0          71.961143  
1          68.182268  
2          68.723056  
3                NaN  
4          69.127688

In [30]:
modified_rank_df.describe().T

count          mean            std          min  \
farmer                  40.0    330.825000     208.216220    14.000000   
avg_time_to_grn         40.0     49.191338       5.201290    48.000000   
total_actual_quantity   40.0  78276.350000  214520.244455  1389.000000   
avg_quantity_variance   40.0    644.554685     802.969308    11.000000   
indent_count            40.0      3.275000       4.782380     1.000000   
lots_supplied           40.0      5.775000      10.216848     1.000000   
avg_grading_yield       39.0     67.409046       7.883984    36.611912   

                               25%          50%           75%           max  
farmer                  121.500000    367.00000    524.250000  5.740000e+02  
avg_time_to_grn          48.000000     48.00000     48.000000  7.200000e+01  
total_actual_quantity  6059.750000  17747.00000  42112.250000  1.200504e+06  
avg_quantity_variance   241.700000    419.05770    663.205000  4.519500e+03  
indent_count              1.000000      2.00000      3.000000  2.200000e+01  
lots_supplied             2.000000      3.00000      4.000000  5.100000e+01  
avg_grading_yield        67.536264     69.27447     70.323846  7.621201e+01

In [31]:
modified_rank_df.head(50)

farmer               farmer_name  avg_time_to_grn  total_actual_quantity  \
0       14            BRAHMA NAIDU K          48.0000                92176.0   
1       18                   K SRINU          48.0000                 5867.0   
2       21                   KS RAJU          48.0000                 4741.0   
3       32      RAVINDRA ENTERPRISES          48.0000                 2176.0   
4       34  SRI KANAKA DURGA TRADERS          48.0000                82161.0   
5       39        CHANDRIKA SEAFOODS          48.0000                37878.0   
6       40            SRI RAMA AQUA           48.0000                 6056.0   
7       47    SRAVANTHI AQUA TRADERS          48.0000                13484.0   
8       82                   SV AQUA          48.0000                12806.0   
9      105           SYAMALA TRADERS          48.0000                19928.0   
10     127                   RAMBABU          48.0000                24949.0   
11     169             RAMALINGARAJU          48.0000                19354.0   
12     177          CH NAGESWARA RAO          48.0000                 8220.0   
13     211         BHARGAVI SEAFOODS          48.0000                 6650.0   
14     267         T SHYAM SUNDHARAM          48.0000                 6272.0   
15     268           ARUNODAYA AQUA           48.0000                28401.0   
16     301         SRINIVASA TRADERS          48.0000                16140.0   
17     322           HIMAJA SEAFOODS          48.0000                 2895.0   
18     338         SRIKANTH SEAFOODS          48.0000                25044.0   
19     355            G CHANDRAMOHAN          48.0000                25271.0   
20     379                LORAL AQUA          48.0000                 5758.0   
21     384     SAI RAMA FROZEN FOODS          48.0000                 6061.0   
22     437      LAUREL AQUA AGENCIES          48.0000               161483.0   
23     454             PV TRADELINKS          48.5106               119684.0   
24     468                   AVINASH          48.0000                10884.0   
25     470  SREE KUMAR FOOD PRODUCTS          48.0000                32952.0   
26     503          UBR AQUA TRADING          71.1429                54815.0   
27     507        VELANKKANI TRADERS          48.0000              1200504.0   
28     512                  SAI AQUA          48.0000                 1389.0   
29     520                  JR FOODS          48.0000                72846.0   
30     537   CHANDRA MOHAN SEA FOODS          48.0000                29601.0   
31     543                  VIKRANTH          48.0000                 8972.0   
32     560         LAVANYA SEA FOODS          48.0000               703750.0   
33     563           GNV SIVA SANKAR          48.0000                 3588.0   
34     569             SS AQUA NEEDS          48.0000                14273.0   
35     570       LIKITH MARINE FOODS          48.0000                26200.0   
36     571             DSR SEA FOODS          48.0000                 1864.0   
37     572     LAKSHMI SAI SEA FOODS          48.0000                 3553.0   
38     573  ROHINI NEW AQUA AGENCIES          72.0000               157850.0   
39     574          RAMA SAI TRADERS          48.0000                74558.0   

    avg_quantity_variance                                         indent_ids  \
0               1369.6250                                               8197   
1                316.5000                                          8094,8190   
2                259.0000                                               8098   
3                324.0000                                               8208   
4                402.1154                           8125,8149,8156,8191,8202   
5               1003.4118                                     8143,8153,8166   
6                472.0000                                          8175,8220   
7                237.0000                                               8169   
8  

In [32]:
numeric_columns_impact_mapper = {
    "avg_time_to_grn": "-",
    "total_actual_quantity": "+",
    "avg_quantity_variance": "-",
    "lots_supplied": "+",
    "avg_grading_yield": "+",
    "indent_count": "+",
}
farmer_ids = modified_rank_df[["farmer", "farmer_name"]].copy()

numeric_dataset = modified_rank_df[
    ["farmer"] + list(numeric_columns_impact_mapper.keys())
].copy()
numeric_dataset = numeric_dataset.fillna(numeric_dataset.mean())

nCol = numeric_dataset.shape[1]

impact = list(numeric_columns_impact_mapper.values())  # Adjusted for 6 numeric columns
equal_weight_factor = 1 / (nCol - 1)  # Equal weight factor for normalization
weights = [equal_weight_factor] * (nCol - 1)

ranked_farmers = rank_farmers(numeric_dataset, nCol, impact, weights)

# Merge with original dataset to get all columns, then add ranking information
final_results = modified_rank_df.merge(
    ranked_farmers[["farmer", "relative_closeness", "rank"]], on="farmer", how="left"
)
final_results = final_results.sort_values(by="rank")
final_results.head()

/tmp/ipykernel_165201/2571334127.py:12: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.00906798706171461' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  dataset.iloc[j, i] = (dataset.iloc[j, i] / temp) * weights[i - 1]
/tmp/ipykernel_165201/2571334127.py:12: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.004585612375822924' has dtype incompatible with int64, please explicitly cast to a compatible dtype first.
  dataset.iloc[j, i] = (dataset.iloc[j, i] / temp) * weights[i - 1]


farmer               farmer_name  avg_time_to_grn  total_actual_quantity  \
27     507        VELANKKANI TRADERS          48.0000              1200504.0   
32     560         LAVANYA SEA FOODS          48.0000               703750.0   
23     454             PV TRADELINKS          48.5106               119684.0   
39     574          RAMA SAI TRADERS          48.0000                74558.0   
4       34  SRI KANAKA DURGA TRADERS          48.0000                82161.0   

    avg_quantity_variance                                         indent_ids  \
27              1431.0138  8095,8096,8099,8102,8103,8110,8111,8119,8128,8...   
32               565.9344  8101,8132,8142,8157,8158,8161,8162,8164,8168,8...   
23               916.3830  8097,8100,8114,8121,8122,8130,8134,8137,8150,8...   
39               268.3750       8171,8176,8178,8179,8180,8196,8210,8211,8221   
4                402.1154                           8125,8149,8156,8191,8202   

    indent_count                                  plant_lot_numbers  \
27            22  5P1-432,5P1-433,5P1-434,5P1-435,5P1-441,5P1-44...   
32            20  5P1-482,5P1-483,5P1-498,5P1-499,5P1-500,5P1-50...   
23            14  5P1-472,5P1-489,5P1-527,5P2-486,5P2-492,5P2-50...   
39             9  5P1-532,5P2-564,5P2-570,5P2-571,5P2-572,5P2-57...   
4              5  5P1-466,5P1-487,5P1-488,5P1-496,5P1-497,5P1-55...   

    lots_supplied  avg_grading_yield  relative_closeness  rank  
27             51          68.660438            0.856856   1.0  
32             44          68.967260            0.748241   2.0  
23             17          70.271716            0.434537   3.0  
39             10          70.765228            0.409543   4.0  
4               8          69.127688            0.379850   5.0

In [33]:
final_results.to_excel("modified_farmer_performance_ranking.xlsx", index=False)

In [34]:
query3 = """
WITH farmer_service AS (
    SELECT 
        ii.farmer,
        MIN(g.created_at) AS first_grn_date,
        TIMESTAMPDIFF(MONTH, MIN(g.created_at), CURDATE()) AS farmer_service_period_in_months
    FROM erpx_dev_rm_procurement.indent_items AS ii
    JOIN erpx_dev_rm_procurement.indent AS i ON ii.indent_id = i.indent_id
    LEFT JOIN erpx_dev_rm_procurement.grn AS g ON g.indent_no = i.indent_id
    WHERE g.created_at IS NOT NULL
    GROUP BY ii.farmer
),
hl_cte AS (
    SELECT
        gi.grn_id,
        gi.created_at AS grn_created_at,
        pg.lot_name,
        gi.count AS hon_count,
        gi.quantity AS hon_weight,
        pg.created_at AS grading_created_at,
        FLOOR(SUM(pgr.weight) - (pg.crate_weight * COUNT(pgr.crates))) AS hl_weight,
        ( (SUM(pgr.weight) - (pg.crate_weight * COUNT(pgr.crates))) / NULLIF(gi.quantity, 0) ) * 100 AS grading_yield,
        GROUP_CONCAT(DISTINCT pgr.sale_order ORDER BY pgr.sale_order ASC SEPARATOR ', ') AS grading_sale_orders
    FROM erpx_dev_rm_procurement.grn_items gi
    LEFT JOIN erpx_dev_production.pp_grading pg
        ON gi.plant_lot_number = pg.lot_name
    LEFT JOIN erpx_dev_production.pp_grading_readings pgr
        ON pg.session_id = pgr.session_id
    GROUP BY 
        gi.grn_id,
        gi.created_at,
        pg.lot_name,
        gi.count,
        gi.quantity,
        pg.created_at
)
SELECT
    ii.farmer,
    s.name AS farmer_name,
    fl.first_grn_date,
    fl.farmer_service_period_in_months,

    AVG(TIMESTAMPDIFF(HOUR, i.indent_date, g.grn_date)) AS avg_time_to_grn,

    SUM(gi.quantity) AS total_actual_quantity,

    AVG(ABS(gi.quantity - ii.expected_qty)) AS avg_quantity_variance,
    
    GROUP_CONCAT(DISTINCT i.indent_id) AS indent_ids,
    COUNT(DISTINCT i.indent_id) AS indent_count,
    
    GROUP_CONCAT(DISTINCT gi.plant_lot_number) AS plant_lot_numbers,
    COUNT(DISTINCT gi.plant_lot_number) AS lots_supplied,
    
    AVG(hl.grading_yield) AS avg_grading_yield

FROM
    erpx_dev_rm_procurement.indent_items AS ii
JOIN
    erpx_dev_rm_procurement.indent AS i ON ii.indent_id = i.indent_id
LEFT JOIN
    erpx_dev_rm_procurement.grn AS g ON g.indent_no = i.indent_id
LEFT JOIN 
    erpx_dev_rm_procurement.grn_items AS gi ON gi.grn_id = g.grn_id
LEFT JOIN
    erpx_dev.suppliers AS s ON s.supplier_id = ii.farmer
LEFT JOIN 
    hl_cte AS hl ON hl.grn_id = gi.grn_id
LEFT JOIN
    farmer_service AS fl ON fl.farmer = ii.farmer
WHERE
    g.grn_date >= DATE_SUB(CURDATE(), INTERVAL 1 MONTH)
GROUP BY
    ii.farmer
ORDER BY
    ii.farmer;
"""

third_modified_rank_df = pd.read_sql(query3, engine)
third_modified_rank_df.head(20)

farmer               farmer_name             first_grn_date  \
0       14            BRAHMA NAIDU K 2023-01-24 15:08:12.917935   
1       18                   K SRINU 2023-01-24 15:40:34.713129   
2       21                   KS RAJU 2023-01-24 16:19:12.397165   
3       32      RAVINDRA ENTERPRISES 2023-01-27 15:51:10.478166   
4       34  SRI KANAKA DURGA TRADERS 2023-01-27 15:45:25.114539   
5       39        CHANDRIKA SEAFOODS 2023-01-27 16:43:15.666488   
6       40            SRI RAMA AQUA  2023-01-27 16:44:40.708132   
7       47    SRAVANTHI AQUA TRADERS 2023-01-30 15:04:30.072026   
8       82                   SV AQUA 2023-02-07 14:39:22.584317   
9      105           SYAMALA TRADERS 2023-02-18 17:33:53.597636   
10     127                   RAMBABU 2023-02-22 15:44:59.255379   
11     169             RAMALINGARAJU 2023-03-02 12:39:52.978513   
12     177          CH NAGESWARA RAO 2023-03-06 12:39:20.739264   
13     211         BHARGAVI SEAFOODS 2023-04-10 16:39:16.952757   
14     267         T SHYAM SUNDHARAM 2023-05-15 17:06:43.941512   
15     268           ARUNODAYA AQUA  2023-05-16 14:29:20.978744   
16     301         SRINIVASA TRADERS 2023-06-07 14:04:03.674874   
17     322           HIMAJA SEAFOODS 2023-06-27 17:08:41.361327   
18     338         SRIKANTH SEAFOODS 2023-07-22 17:23:38.513517   
19     355            G CHANDRAMOHAN 2023-07-28 10:37:41.815338   

    farmer_service_period_in_months  avg_time_to_grn  total_actual_quantity  \
0                                31             48.0                92176.0   
1                                31             48.0                 5867.0   
2                                31             48.0                 4741.0   
3                                30             48.0                 2176.0   
4                                30             48.0                82161.0   
5                                30             48.0                37878.0   
6                                30             48.0                 6056.0   
7                                30             48.0                13484.0   
8                                30             48.0                12806.0   
9                                30             48.0                19928.0   
10                               30             48.0                24949.0   
11                               29             48.0                19354.0   
12                               29             48.0                 8220.0   
13                               28             48.0                 6650.0   
14                               27             48.0                 6272.0   
15                               27             48.0                28401.0   
16                               26             48.0                16140.0   
17                               25             48.0                 2895.0   
18                               25             48.0                25044.0   
19                               24             48.0                25271.0   

    avg_quantity_variance                indent_ids  indent_count  \
0               1369.6250                      8197             1   
1                316.5000                 8094,8190             2   
2                259.0000                      8098             1   
3                324.0000                      8208             1   
4                402.1154  8125,8149,8156,8191,8202             5   
5               1003.4118            8143,8153,8166             3   
6                472.0000                 8175,8220             2   
7                237.0000                      8169             1   
8                219.2000            8141,8212,8219             3   
9                240.8000            8163,8185,8203             3   
10               150.0357                 8104,8118             2   
11               596.6667                 8120,8160             2   
12              4519.5000              

In [78]:
query4 = """
WITH farmer_service AS (
    SELECT 
        ii.farmer,
        MIN(g.created_at) AS first_grn_date,
        TIMESTAMPDIFF(MONTH, MIN(g.created_at), CURDATE()) AS farmer_service_period_in_months
    FROM erpx_dev_rm_procurement.indent_items AS ii
    JOIN erpx_dev_rm_procurement.indent AS i ON ii.indent_id = i.indent_id
    LEFT JOIN erpx_dev_rm_procurement.grn AS g ON g.indent_no = i.indent_id
    WHERE g.created_at IS NOT NULL
    GROUP BY ii.farmer
),
hl_cte AS (
    SELECT
        g.grn_id,
        g.grn_item_id,
        g.created_at GRN_Created_dt,
        p.lot_name,
        g.count as hon_count,
        g.quantity as hon_weight,
        p.created_at as Grading_Created_dt,
        FLOOR(SUM(v.weight) - (p.crate_weight * COUNT(v.crates))) AS hl_weight,
        ((SUM(v.weight) - (p.crate_weight * COUNT(v.crates))) / g.quantity) * 100 AS grading_yield,
        GROUP_CONCAT(DISTINCT v.sale_order ORDER BY v.sale_order ASC SEPARATOR ', ') AS grading_sale_orders
    FROM
        erpx_dev_rm_procurement.grn_items g
    LEFT JOIN erpx_dev_production.pp_grading p
        ON g.plant_lot_number = p.lot_name
    LEFT JOIN erpx_dev_production.pp_grading_readings v
        ON p.session_id = v.session_id
    -- WHERE g.created_at >= DATE_SUB(CURDATE(), INTERVAL 2 MONTH)
    GROUP BY
        g.grn_item_id,
        g.created_at,
        p.lot_name,
        g.count,
        g.quantity,
        p.created_at
    ),
grading_count_cte AS (
    SELECT
        lot_name,
        SUM(net_grading_weight * grading_count * 2.20462) / SUM(net_grading_weight) AS weighted_avg_shrimp_count_per_kg
    FROM (
        SELECT
            pg.lot_name,
            pgr.count AS grading_count,
            SUM(pgr.weight) - (pg.crate_weight * SUM(pgr.crates)) AS net_grading_weight
        FROM pp_grading pg
        LEFT JOIN pp_grading_readings pgr 
            ON pg.session_id = pgr.session_id
        GROUP BY pg.session_id, pg.lot_name
    ) t
    GROUP BY lot_name
)
SELECT
    ii.farmer,
    s.name AS farmer_name,
    fl.first_grn_date,
    fl.farmer_service_period_in_months,

    AVG(TIMESTAMPDIFF(HOUR, i.indent_date, g.grn_date)) AS avg_time_to_grn,
    
    SUM(gi.quantity) AS total_actual_quantity,

    AVG(ABS(gi.quantity - ii.expected_qty)) AS avg_quantity_variance,
    AVG(ABS(gi.count - ii.expected_count)) AS abs_avg_count_variance_indent_grn,

    AVG(gc.weighted_avg_shrimp_count_per_kg) AS avg_shrimp_count_at_grading,
    AVG(gc.weighted_avg_shrimp_count_per_kg - gi.count) AS avg_count_variance_grn_grading,

    -- GROUP_CONCAT(DISTINCT i.indent_id) AS indent_ids,
    COUNT(DISTINCT i.indent_id) AS indent_count,
    
    GROUP_CONCAT(DISTINCT gi.plant_lot_number) AS plant_lot_numbers,
    COUNT(DISTINCT gi.plant_lot_number) AS lots_supplied,
    
    AVG(hl.grading_yield) AS avg_grading_yield

FROM
    erpx_dev_rm_procurement.indent_items AS ii
LEFT JOIN
	erpx_dev_rm_procurement.grn_items AS gi ON gi.indent_item_id = ii.indent_item_id
LEFT JOIN
	erpx_dev_rm_procurement.indent AS i ON ii.indent_id = i.indent_id
LEFT JOIN
    erpx_dev_rm_procurement.grn AS g ON g.grn_id  = gi.grn_id 
LEFT JOIN
    erpx_dev.suppliers AS s ON s.supplier_id = ii.farmer
LEFT JOIN 
    hl_cte AS hl ON hl.grn_item_id = gi.grn_item_id
LEFT JOIN
    farmer_service AS fl ON fl.farmer = ii.farmer
LEFT JOIN
    grading_count_cte AS gc ON gc.lot_name = gi.plant_lot_number
WHERE
    g.grn_date >= DATE_SUB(CURDATE(), INTERVAL 24 MONTH)
GROUP BY
    ii.farmer
ORDER BY
    ii.farmer;
"""

fourth_modified_rank_df = pd.read_sql(query4, engine)
fourth_modified_rank_df.head(20)

farmer              farmer_name             first_grn_date  \
0        1              Karuna Raju 2023-01-20 18:43:15.625959   
1        2   Sri Lakshmi Aqua Farms 2023-01-21 16:06:28.127043   
2        4     Varalakshmi Agencies 2023-01-23 14:43:10.667844   
3        8                   K Sriu 2023-01-24 11:14:44.296851   
4       10           Sita rama raju 2023-01-24 14:20:08.997527   
5       12    SRI RAMA AQUA TRADERS 2023-01-24 14:56:07.383992   
6       13      MEENAKSHI FISHEREIS 2023-01-24 15:02:18.693342   
7       14           BRAHMA NAIDU K 2023-01-24 15:08:12.917935   
8       15    VENKATESWARA SEAFOODS 2023-01-24 15:14:28.265919   
9       17         P SITA RAMA RAJU 2023-01-24 15:30:08.480028   
10      18                  K SRINU 2023-01-24 15:40:34.713129   
11      19           SRINIVASARAO K 2023-01-24 16:04:20.257398   
12      20  SAI LAKSHMI AUQA FARMS  2023-01-24 16:12:41.923277   
13      21                  KS RAJU 2023-01-24 16:19:12.397165   
14      22             SKM SEAFOODS 2023-01-24 16:23:44.257040   
15      24                    rahul 2023-01-24 19:42:46.849846   
16      25          G V PHANI KUMAR 2023-01-25 10:56:31.293553   
17      26            SWAMY NAIDU K 2023-01-25 11:24:21.853013   
18      27   GOPALA KRISHNAM RAJU K 2023-01-27 16:11:32.140816   
19      28         SOMESWARA RAJU D 2023-01-27 16:11:16.485284   

    farmer_service_period_in_months  avg_time_to_grn  total_actual_quantity  \
0                                31           0.0000                 9325.0   
1                                31          48.0000                 3584.0   
2                                31          48.0000                 2887.0   
3                                31           0.0000                 2550.0   
4                                31          59.0645               142653.0   
5                                31          96.0000                 1090.0   
6                                31         148.0952                94589.0   
7                                31          65.4545                18039.0   
8                                31          64.5246               139807.0   
9                                31          64.6667                 7517.0   
10                               31          60.5674               379360.0   
11                               31          58.5467              1183636.0   
12                               31          70.8966                74716.0   
13                               31          59.6771              1171301.0   
14                               31         205.9643                56382.0   
15                               31         598.0000                 2010.0   
16                               31          48.0000                 1912.0   
17                               31          72.0000                11222.0   
18                               30          70.6667                 9931.0   
19                               30          88.0000                45616.0   

    avg_quantity_variance  abs_avg_count_variance_indent_grn  \
0                  0.0000                             0.0000   
1                116.0000                             5.0000   
2                 56.5000                             5.0000   
3                 25.0000                             1.0000   
4                350.3548                             6.5161   
5                 90.0000                            10.0000   
6                167.3175                             5.8254   
7                 67.3636                             4.8182   
8                193.0820                             4.0492   
9                 96.3333                            10.3333   
10               600.7660                            11.2553   
11               767.3267                             9.4233   
12               381.3793                            14.6207   
13               677.6563                             7.0521 

In [79]:
fourth_modified_rank_df.describe().T

count                           mean  \
farmer                             360.0                     333.772222   
first_grn_date                       360  2023-11-11 08:49:57.026438400   
farmer_service_period_in_months    360.0                          20.95   
avg_time_to_grn                    360.0                     106.241015   
total_actual_quantity              360.0                   58700.561111   
avg_quantity_variance              360.0                     214.916727   
abs_avg_count_variance_indent_grn  360.0                       9.131003   
avg_shrimp_count_at_grading        121.0                      75.629299   
avg_count_variance_grn_grading     121.0                      24.480896   
indent_count                       360.0                      11.169444   
lots_supplied                      360.0                      20.319444   
avg_grading_yield                  121.0                      81.216881   

                                                          min  \
farmer                                                    1.0   
first_grn_date                     2023-01-20 18:43:15.625959   
farmer_service_period_in_months                           0.0   
avg_time_to_grn                                           0.0   
total_actual_quantity                                   133.0   
avg_quantity_variance                                     0.0   
abs_avg_count_variance_indent_grn                         0.0   
avg_shrimp_count_at_grading                         25.904285   
avg_count_variance_grn_grading                       -13.8614   
indent_count                                              1.0   
lots_supplied                                             1.0   
avg_grading_yield                                   15.379601   

                                                             25%  \
farmer                                                     176.5   
first_grn_date                     2023-03-08 08:10:19.249478912   
farmer_service_period_in_months                             15.0   
avg_time_to_grn                                             48.0   
total_actual_quantity                                     4690.0   
avg_quantity_variance                                       64.5   
abs_avg_count_variance_indent_grn                            5.0   
avg_shrimp_count_at_grading                             50.70626   
avg_count_variance_grn_grading                         12.167913   
indent_count                                                 1.0   
lots_supplied                                                2.0   
avg_grading_yield                                      66.675616   

                                                             50%  \
farmer                                                     389.5   
first_grn_date                     2023-09-07 17:38:06.653643776   
farmer_service_period_in_months                             23.0   
avg_time_to_grn                                          55.9722   
total_actual_quantity                                    10638.5   
avg_quantity_variance                                   114.4375   
abs_avg_count_variance_indent_grn                            6.0   
avg_shrimp_count_at_grading                            75.631964   
avg_count_variance_grn_grading                         20.087445   
indent_count                                                 2.0   
lots_supplied                                                4.0   
avg_grading_yield                                      69.768432   

                                                             75%  \
farmer                                                    482.25   
first_grn_date                     2024-05-23 04:26:27.638885888   
farmer_service_period_in_months                             29.0   
avg_time_to_grn                                             72.0   
total_actual_quantity                                    38211.0   
avg_quantity_variance              